Rust build exception

Hi! I migrated to the new version with native assets support. However I can't build my app anymore... Looks like the rust compiler is missing in the NDK ? I have pretty much no idea how any of those things work, so any help would be very appreciated 😊

```
Unhandled exception:
RustupBuildException(message: Expected binary C:\Users\maelc\AppData\Local\Android\Sdk\ndk\27.3.13750724\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android35-clang.exe not found; is your installed compiler too old?)
#0      createBuildEnvVars.getBinary (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:242:7)
GregoryConrad/mimir#1      createBuildEnvVars (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:269:33)
GregoryConrad/mimir#2      main.<anonymous closure> (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:70:21)
<asynchronous suspension>
GregoryConrad/mimir#3      build (package:hooks/src/api/build_and_link.dart:105:5)
<asynchronous suspension>
GregoryConrad/mimir#4      main (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:9:3)
<asynchronous suspension>

  Building assets for package:mimir failed.
  build.dart returned with exit code: 255.
  To reproduce run:
  (cd C:\Users\maelc\Documents\dev\Frameworks\pub\Cache\hosted\pub.dev\mimir-0.2.0+1\; C:\Users\maelc\Documents\dev\Frameworks\flutter\bin\cache\dart-sdk\bin\dart --packages=C:\Users\maelc\Documents\dev\Projects\LocalMaterialNotes\.dart_tool\package_config.json C:\Users\maelc\Documents\dev\Projects\LocalMaterialNotes\.dart_tool\hooks_runner\mimir\546a066584db07cdf2a7c014ce9d34ea\hook.dill --config=C:\Users\maelc\Documents\dev\Projects\LocalMaterialNotes\.dart_tool\hooks_runner\mimir\546a066584db07cdf2a7c014ce9d34ea\input.json )
  stderr:
  Unhandled exception:
RustupBuildException(message: Expected binary C:\Users\maelc\AppData\Local\Android\Sdk\ndk\27.3.13750724\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android35-clang.exe not found; is your installed compiler too old?)
#0      createBuildEnvVars.getBinary (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:242:7)
GregoryConrad/mimir#1      createBuildEnvVars (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:269:33)
GregoryConrad/mimir#2      main.<anonymous closure> (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:70:21)
<asynchronous suspension>
GregoryConrad/mimir#3      build (package:hooks/src/api/build_and_link.dart:105:5)
<asynchronous suspension>
GregoryConrad/mimir#4      main (file:///c:/users/maelc/documents/dev/frameworks/pub/cache/hosted/pub.dev/mimir-0.2.0+1/hook/build.dart:9:3)
<asynchronous suspension>

  stdout:
  [CONFIG] mimir_build: {
  "assets": {},
  "config": {
    "build_asset_types": [
      "code_assets/code"
    ],
    "extensions": {
      "code_assets": {
        "android": {
          "target_ndk_api": 24
        },
        "c_compiler": {
          "ar": "C:\\Users\\maelc\\AppData\\Local\\Android\\Sdk\\ndk\\27.3.13750724\\toolchains\\llvm\\prebuilt\\windows-x86_64\\bin\\llvm-ar.exe",
          "cc": "C:\\Users\\maelc\\AppData\\Local\\Android\\Sdk\\ndk\\27.3.13750724\\toolchains\\llvm\\prebuilt\\windows-x86_64\\bin\\clang.exe",
          "ld": "C:\\Users\\maelc\\AppData\\Local\\Android\\Sdk\\ndk\\27.3.13750724\\toolchains\\llvm\\prebuilt\\windows-x86_64\\bin\\ld.lld.exe"
        },
        "link_mode_preference": "dynamic",
        "target_architecture": "x64",
        "target_os": "android"
      }
    },
    "linking_enabled": false
  },
  "out_dir_shared": "C:\\Users\\maelc\\Documents\\dev\\Projects\\LocalMaterialNotes\\.dart_tool\\hooks_runner\\shared\\mimir\\build\\",
  "out_file": "C:\\Users\\maelc\\Documents\\dev\\Projects\\LocalMaterialNotes\\.dart_tool\\hooks_runner\\mimir\\546a066584db07cdf2a7c014ce9d34ea\\output.json",
  "package_name": "mimir",
  "package_root": "C:\\Users\\maelc\\Documents\\dev\\Frameworks\\pub\\Cache\\hosted\\pub.dev\\mimir-0.2.0+1\\",
  "user_defines": {}
}
[CONFIG] mimir_build: {ANDROID_HOME: C:\Users\maelc\AppData\Local\Android\Sdk, COMSPEC: C:\WINDOWS\SYSTEM32\cmd.exe, PATH: C:\Users\maelc\.jdks\jbr-17.0.12\bin;C:\Program Files\Oculus\Support\oculus-runtime;C:\Ruby33-x64\bin;C:\Program Files (x86)\Common Files\Intel\Shared Libraries\redist\intel64\compiler;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\dotnet\;C:\MinGW\bin;C:\ProgramData\chocolatey\bin;C:\Program Files\OpenJDK\openjdk-11.0.13_8\bin;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\NVIDIA Corporation\NVIDIA App\NvDLISR;C:\Users\maelc\Documents\dev\Frameworks\pub\Cache\bin;C:\Users\maelc\Documents\dev\Frameworks\flutter\bin;C:\Program Files\nodejs\;C:\Program Files\FirefoxPWA\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Git\cmd;C:\Program Files\Docker\Docker\resources\bin;C:\Program Files\PowerShell\7\;C:\Program Files\GitHub CLI\;C:\Users\maelc\AppData\Local\Programs\Python\Launcher\;C:\Users\maelc\AppData\Local\Programs\Python\Python313\Scripts\;E:\Documents\Autres\Logiciels\Bitwarden;C:\Users\maelc\AppData\Local\Programs\Python\Python313\;C:\ProgramData\chocolatey\lib\make;C:\Users\maelc\.cargo\bin;C:\Users\maelc\scoop\shims;C:\Users\maelc\AppData\Local\Programs\Python\Python310\Scripts\;C:\Users\maelc\AppData\Local\Programs\Python\Python310\;C:\Users\maelc\AppData\Local\Microsoft\WindowsApps;C:\Users\maelc\AppData\Local\Android\Sdk\platform-tools;C:\Users\maelc\AppData\Local\JetBrains\Toolbox\scripts;C:\Users\maelc\AppData\Local\Programs\oh-my-posh\bin;C:\Users\maelc\.dotnet\tools;C:\Users\maelc\AppData\Local\Android\Sdk\emulator;C:\Users\maelc\AppData\Roaming\npm;C:\Users\maelc\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\maelc\AppData\Local\Microsoft\WindowsApps;C:\Users\maelc\AppData\Local\Programs\Android Studio\jbr\bin;C:\Program Files\SSHFS-Win\bin;C:\Users\maelc\AppData\Local\Microsoft\WinGet\Links;;C:\Users\maelc\Documents\dev\Frameworks\flutter\bin\mingit\cmd;C:\Users\maelc\Documents\dev\Frameworks\flutter\bin\mingit\cmd, PATHEXT: .COM;.EXE;.BAT;.CMD;.VBS;.JS;.WS;.MSC, PROGRAMDATA: C:\ProgramData, PROMPT: $P$G, SYSTEMDRIVE: C:, SYSTEMROOT: C:\WINDOWS, TEMP: C:\Users\maelc\AppData\Local\Temp, TMP: C:\Users\maelc\AppData\Local\Temp, USERPROFILE: C:\Users\maelc, WINDIR: C:\WINDOWS}
[INFO] mimir_build: Invoking "rustup [show, active-toolchain]" in directory C:\Users\maelc\Documents\dev\Frameworks\pub\Cache\hosted\pub.dev\mimir-0.2.0+1\native with environment: {}
[INFO] mimir_build: Invoking "rustup [target, list, --toolchain, 1.90.0-x86_64-pc-windows-msvc]" with environment: {}


Target dart_build failed: Error: Building native assets failed. See the logs for more details.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command 'C:\Users\maelc\Documents\dev\Frameworks\flutter\bin\flutter.bat'' finished with non-zero exit value 1
```

--- a/native_toolchain_rs/lib/src/build_environment.dart
+++ b/native_toolchain_rs/lib/src/build_environment.dart
@@ -11,31 +11,6 @@ interface class BuildEnvironmentFactory {
   const BuildEnvironmentFactory();
 
   Map<String, String> createBuildEnvVars(CodeConfig codeConfig) {
-    final CodeConfig(:targetOS, :cCompiler) = codeConfig;
-    final targetTriple = codeConfig.targetTriple;
-    final targetTripleEnvVar = targetTriple.replaceAll('-', '_');
-
-    String getBinary(String binaryName) {
-      if (cCompiler == null) {
-        throw UnsupportedError(
-          'CCompilerConfig was not provided but is required for $targetTriple',
-        );
-      }
-
-      final binaryPath = path.join(
-        path.dirname(path.fromUri(cCompiler.compiler)),
-        OS.current.executableFileName(binaryName),
-      );
-
-      if (!File(binaryPath).existsSync()) {
-        throw RustValidationException([
-          'Binary $binaryPath not found; is your installed compiler too old?',
-        ]);
-      }
-
-      return binaryPath;
-    }
-
     return {
       // NOTE: XCode makes some injections into PATH that break host build
       // for crates with a build.rs
@@ -47,19 +22,65 @@ interface class BuildEnvironmentFactory {
             .join(':'),
       },
 
-      // NOTE: we need to point to NDK >=27 vended LLVM for Android.
-      // The `${targetTriple}35-clang`s were introduced in NDK 27,
-      // so using these binaries:
-      // 1. Ensures we are using a compatible NDK
-      // 2. Also fixes build issues when just using the `clang`s directly
-      if (targetOS == OS.android) ...{
-        'AR_$targetTripleEnvVar': getBinary('llvm-ar'),
-        'CC_$targetTripleEnvVar': getBinary('${targetTriple}35-clang'),
-        'CXX_$targetTripleEnvVar': getBinary('${targetTriple}35-clang++'),
-        'CARGO_TARGET_${targetTripleEnvVar.toUpperCase()}_LINKER': getBinary(
-          '${targetTriple}35-clang',
+      if (codeConfig.targetOS == OS.android)
+        ...const AndroidBuildEnvironmentFactory().createBuildEnvVars(
+          codeConfig: codeConfig,
         ),
-      },
+    };
+  }
+}
+
+@internal
+interface class AndroidBuildEnvironmentFactory {
+  const AndroidBuildEnvironmentFactory();
+
+  Map<String, String> createBuildEnvVars({
+    required CodeConfig codeConfig,
+  }) {
+    final CodeConfig(:targetTriple, :cCompiler) = codeConfig;
+    if (cCompiler == null) {
+      throw UnsupportedError(
+        'CCompilerConfig was not provided but is required for $targetTriple',
+      );
+    }
+
+    String getCompilerPath(String binaryName) {
+      final compilerBinariesDir = path.dirname(
+        path.fromUri(cCompiler.compiler),
+      );
+      final binaryPath = path.join(compilerBinariesDir, binaryName);
+
+      if (!File(binaryPath).existsSync()) {
+        throw RustValidationException([
+          'Binary $binaryPath not found; is your installed NDK too old?',
+        ]);
+      }
+
+      return binaryPath;
+    }
+
+    final targetTripleEnvVar = targetTriple.replaceAll('-', '_');
+    final ndkTargetTriple = switch (targetTriple) {
+      // NOTE: sometimes the Rust and NDK target triples do not match.
+      // See: https://github.com/GregoryConrad/native_toolchain_rs/issues/21#issuecomment-3368307228
+      'armv7-linux-androideabi' => 'armv7a-linux-androideabi',
+      _ => targetTriple,
+    };
+
+    // NOTE: we need to point to NDK >=27 vended LLVM for Android.
+    // The `${ndkTargetTriple}35-clang`s were introduced in NDK 27,
+    // so using these binaries:
+    // 1. Ensures we are using a compatible NDK
+    // 2. Also fixes build issues when just using the `clang`s directly
+    const apiTarget = '35';
+    final clangPath = getCompilerPath('$ndkTargetTriple$apiTarget-clang');
+    final clangPpPath = getCompilerPath('$ndkTargetTriple$apiTarget-clang++');
+
+    return {
+      'AR_$targetTripleEnvVar': path.fromUri(cCompiler.archiver),
+      'CC_$targetTripleEnvVar': clangPath,
+      'CXX_$targetTripleEnvVar': clangPpPath,
+      'CARGO_TARGET_${targetTripleEnvVar.toUpperCase()}_LINKER': clangPath,
     };
   }
 }

